### ISSUE TITLE
Greengrass V2 interprocess communication (IPC)

### ISSUE THREAD

**Issue by fiLLLip** [aws/aws-iot-device-sdk-js-v2#128]
**Is your feature request related to a problem? Please describe.**
We are developing a custom component for Greengrass V2, and need (I think we need this at least) IPC to subscribe to IoT Core via the local service. As far as I can tell, there is no examples nor support in js (this lib) for it now and need to be developed.

**Describe the solution you'd like**
I would like to have the same interfaces and opportunities as in the python version of this SDK. And also updated docs on https://docs.aws.amazon.com/greengrass/v2/developerguide/interprocess-communication.html

**Describe alternatives you've considered**
I have considered connecting directly with IoT Core, but it kind of goes against the "having Greengrass and that handles everything" way of doing it.

**Comment by QuantumLove** [comment]
We have the same problem, Java and Python already support this, but it is not possible for components running Node

**Comment by QuantumLove** [comment]
@justinboswell this is a big stone in our software stack, no thoughts on implementing this yet?

**Comment by skulljoi** [comment]
Any news?

**Comment by jmklix** [comment]
Thanks for the feedback. I have forwarded it to the proper team, but I can't give any promises for when it will be added.

**Comment by samlaf** [comment]
Same here. greengrass ipc support for javascript would be really helpful. Any news yet?

**Comment by charlieharris1** [comment]
I'd really appreciate an update on this. I should have checked IPC was supported by the SDK before spending time on the component I've been working on that's written in JS 🤦 .

**Comment by johnz** [comment]
any update on this, please?

**Comment by jmklix** [comment]
I understand that this is important to you and this has been added to the greengrass teams input queue. But there is no update on this yet. Any feedback on this is appreciated and will help it be prioritized correctly. Make sure to watch this issue for any further updates.

**Comment by olivierpichon** [comment]
IPC is the main way we could use to transform the data which is coming to the local MQTT broker before it gets send to IoT Core. It is not a nice experience to have to swap language only for this. Would be good to prioritize this. Thx!

**Comment by teuteuguy** [comment]
Any updates on this ?

**Comment by AdamVD** [comment]
I'll add my use-case to the pot. We have Node-based API services that need to be deployed in the cloud and on a Greengrass core. The API services need to interact with device shadows in both deployments. In the cloud that's no problem but without support for IPC or shadow operations in this SDK we're going to have to find a workaround. We already planned to use RabbitMQ as our messaging broker between custom components, so we may build some kind of custom bridge component in a supported language to forward shadow requests from our Node services. Not ideal.

**Comment by charlieharris1** [comment]
It looks like IPC support isn't on the horizon so I was wondering if anyone else has deployed a JS component that can publish to IoT core via MQTT and also subscribe to messages. Unless there's something I'm missing it isn't really discussed in the docs.

**Comment by mpetito-envative** [comment]
It seems like one approach could be to connect to the local MQTT broker on the device from JS, but it's unclear how auth should be handled in that scenario and whether that would make your JS component appear as a separate device to IoT Core.

The [MQTT Bridge](https://docs.aws.amazon.com/greengrass/v2/developerguide/mqtt-bridge-component.html) component could then be used to route messages between Local MQTT, PubSub IPC and IoT Core.

**Comment by olivierpichon** [comment]
> It looks like IPC support isn't on the horizon so I was wondering if anyone else has deployed a JS component that can publish to IoT core via MQTT and also subscribe to messages. Unless there's something I'm missing it isn't really discussed in the docs.

@charlieharris1 You could use a mqtt bridge configuration that looks like this:

```json
{
	"mqttTopicMapping": {
		"IotCoreMapping": {
			"topic": "local/+/+/+",
			"source": "LocalMqtt",
			"target": "Pubsub"
		},
		"PubsubMapping": {
			"topic": "dt/+/+/+/+",
			"source": "LocalMqtt",
			"target": "IotCore"
		}
	}
}
```

Then you would have the first device send its data to `local/+/+/+` for example. A second device (or component running on your code device with dedicated credentials) would connect to the local MQTT broker and subscribe to `local/+/+/+`. It could then make any data transformation that needs to happen before sending the data back to IoT core, publishing on `dt/+/+/+/+`.

The real benefit of having IPC supported by the nodejs SDK would be not to have to issue dedicated certificates for this Greengrass component.

**Comment by mpetito-envative** [comment]
In case it helps someone else, we decided to go with this approach and use a local MQTT broker.  We then re-configured the local broker to use Mosquitto without authentication.  Now the bridge component can relay messages to/from IoT Core and local components do not require additional certificates to connect with the local broker.

**Comment by MinCheTsai** [comment]
We want use many greengrass v2 components(ex: shadow manager), but it is through IPC to communication, that mean we can't use node.js to develop our edge application. so please support IPC ASAP.

**Comment by jgutenplan** [comment]
Any updates on whether IPC support is going to be added to node sdk?

**Comment by jgutenplan** [comment]
> In case it helps someone else, we decided to go with this approach and use a local MQTT broker. We then re-configured the local broker to use Mosquitto without authentication. Now the bridge component can relay messages to/from IoT Core and local components do not require additional certificates to connect with the local broker.

@mpetito-envative Can you share a bit more on how you were able to make the Moquette component accept connections without auth? Are you creating your own broker separate from the AWS provided component? I have not been able to get any mqtt clients to connect from a component/lambda to the AWS provided mqtt components.

**Comment by mpetito-envative** [comment]
We didn't use the Moquette component at all.  Instead, we installed a local instance of Mosquitto and reconfigured the Greengrass MQTT Bridge component brokerUri to use it.

This gave us full control over the local broker so that we could configure Mosquitto to accept local connections from other components without authentication.

**Comment by chester-lee1321** [comment]
I'm waiting to be released this feature too.

**Comment by judedaryl** [comment]
@mpetito-envative can you share your bridge config? i have installed a custom mqtt broker and pointed the bridge to it, but it doesn't forward my messages to IoT core


```json
   {
    "mqttTopicMapping": {
      "test": {
        "topic": "clients/asd/hello/worlds",
        "source": "LocalMqtt",
        "target": "IotCore"
      },
      "test2": {
        "topic": "clients/asd/hello/worlds2",
        "source": "IotCore",
        "target": "LocalMqtt"
      }
    },
    "brokerUri": "tcp://localhost:1883"
  }
```

**Comment by bretambrose** [comment]
Greengrass IPC has been released in developer preview in [v1.12.0](https://github.com/aws/aws-iot-device-sdk-js-v2/releases/tag/v1.12.0)
